### PR TITLE
Add nonce to OIDC Authentication Request

### PR DESCRIPTION
**Pull request by mkheck**
This is a PR for gh-4442: Consider adding nonce to OAuth2 Authentication Request (https://github.com/spring-projects/spring-security/issues/4442).

I've coded the implementation & tests for creating+passing+verifying the nonce in the ID token, and all tests pass (with the exception of the one that wasn't passing in the codebase before).

I would like to open a follow-on request to push some optimizing changes out a bit further.
I'm getting everything I need within the `OidcIdTokenValidator` by retrieving the nonce via claims, but it would be cleaner to change from dealing with `Jwt` to dealing with `OIdcIdToken`s instead. But that would necessitate reaching into & reworking `DefaultOidcIdTokenValidatorFactory` (& of course tests associated, etc.) and everything that touches. And it touches quite a lot.

In short, this all works, so I believe the issue to be feature complete. But I would like to extend further changes to present a more specific solution throughout.

### CODE REVIEW

**Comment by jgrandja** [comment]
@mkheck I've been thinking about whether the `nonce` validation should happen in `OidcIdTokenValidator` or `OidcAuthorizationCodeAuthenticationProvider`. Although it seems natural that it should happen in `OidcIdTokenValidator`, there are some challenges with that since either `nonce` should be passed in via the constructor, as you have it:

`public OidcIdTokenValidator(ClientRegistration clientRegistration, String nonce)`

which is not ideal since we would have to create a new `OidcIdTokenValidator` per request since the specific `nonce` is request specific.

The other solution is to implement the validation check in `OidcAuthorizationCodeAuthenticationProvider`, which would work well since we have access to the `OidcIdToken` and `OAuth2AuthorizationRequest`. We are also doing validation checks there between the `OAuth2AuthorizationRequest` and `OAuth2AuthorizationResponse`. So I say we implement the `nonce` check directly in `OidcAuthorizationCodeAuthenticationProvider`.

Makes sense?

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizationCodeAuthenticationProvider.java [hunk 1]
```diff
@@ -211,4 +232,10 @@ private OidcIdToken createOidcToken(ClientRegistration clientRegistration, OAuth
 		OidcIdToken idToken = new OidcIdToken(jwt.getTokenValue(), jwt.getIssuedAt(), jwt.getExpiresAt(), jwt.getClaims());
 		return idToken;
 	}
+
+	private String createHash(String codeVerifier) throws NoSuchAlgorithmException {
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizationCodeAuthenticationProvider.java]: Rename parameter `codeVerifier` to `nonce`

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizationCodeAuthenticationProvider.java [hunk 2]
```diff
@@ -154,6 +162,19 @@ public Authentication authenticate(Authentication authentication) throws Authent
 		}
 		OidcIdToken idToken = createOidcToken(clientRegistration, accessTokenResponse);
 
+		String nonceHash;
+		try {
+			nonceHash = createHash(authorizationRequest.getAttribute(IdTokenClaimNames.NONCE));
+		} catch (NoSuchAlgorithmException e) {
+			OAuth2Error nonceVerificationError = new OAuth2Error(UNVERIFIABLE_NONCE, e.getMessage(), null);
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizationCodeAuthenticationProvider.java]: We should not give too much information to the caller letting them know that the `nonce` could not be verified. A generic message is sufficient so let's reuse `INVALID_NONCE` and remove `UNVERIFIABLE_NONCE `

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizationCodeAuthenticationProvider.java [hunk 3]
```diff
@@ -154,6 +162,19 @@ public Authentication authenticate(Authentication authentication) throws Authent
...
+			throw new OAuth2AuthenticationException(nonceVerificationError);
+		}
+		String nonceClaim = idToken.getClaim(IdTokenClaimNames.NONCE);
+		if (nonceClaim == null || !nonceClaim.equals(nonceHash)) {
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizationCodeAuthenticationProvider.java]: The `nonceClaim` is actually the `nonce` hash so let's rename to `nonceHashClaim`

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenValidator.java [hunk 4]
```diff
@@ -112,7 +112,7 @@ public OAuth2TokenValidatorResult validate(Jwt idToken) {
 		// that it is the same value as the one that was sent in the Authentication Request.
 		// The Client SHOULD check the nonce value for replay attacks.
 		// The precise method for detecting replay attacks is Client specific.
-		// TODO Depends on gh-4442
+		// gh-4442 implemented elsewhere
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenValidator.java]: Let's remove the notes on `nonce`

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java [hunk 5]
```diff
@@ -61,7 +63,7 @@
 	private final ClientRegistrationRepository clientRegistrationRepository;
 	private final AntPathRequestMatcher authorizationRequestMatcher;
 	private final StringKeyGenerator stateGenerator = new Base64StringKeyGenerator(Base64.getUrlEncoder());
-	private final StringKeyGenerator codeVerifierGenerator = new Base64StringKeyGenerator(Base64.getUrlEncoder().withoutPadding(), 96);
+	private final StringKeyGenerator randomKeyGenerator = new Base64StringKeyGenerator(Base64.getUrlEncoder().withoutPadding(), 96);
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java]: Rename member to `stringKeyGenerator`

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java [hunk 6]
```diff
@@ -201,6 +207,28 @@ private static String expandRedirectUri(HttpServletRequest request, ClientRegist
...
+	 * Creates nonce and its hash for use in OpenID Connect Authentication Requests
+	 *
+	 * @param attributes where {@link IdTokenClaimNames#NONCE} is stored for the token request
+	 * @param additionalParameters where hash of {@link IdTokenClaimNames#NONCE} is added to the authentication request
+	 *
+	 * @since 5.2
+	 * @see <a target="_blank" href="https://openid.net/specs/openid-connect-core-1_0.html#NonceNotes">15.5.2.  Nonce Implementation Notes</a>
+	 * @see <a target="_blank" href="https://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation">3.1.3.7.  ID Token Validation</a>
+	 */
+	private void createNonceAndHashForRequest(Map<String, Object> attributes, Map<String, Object> additionalParameters) {
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java]: Please rename to `addNonceParameters`

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java [hunk 7]
```diff
@@ -201,6 +207,28 @@ private static String expandRedirectUri(HttpServletRequest request, ClientRegist
...
+		String nonce = this.randomKeyGenerator.generateKey();
+		attributes.put(IdTokenClaimNames.NONCE, nonce);
+
+		try {
+			String nonceHash = createHash(nonce);
+			additionalParameters.put(IdTokenClaimNames.NONCE, nonceHash);
+		} catch (NoSuchAlgorithmException e) {
+			// MAH: TODO...but what?
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java]: If the `nonce` hash could not be created because of JCE crypto lib missing than we should not add the `nonce` to `attributes`, effectively creating an Authentication Request without a `nonce` as is the current impl.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java [hunk 8]
```diff
@@ -214,18 +242,18 @@ private static String expandRedirectUri(HttpServletRequest request, ClientRegist
...
+			String codeChallenge = createHash(codeVerifier);
 			additionalParameters.put(PkceParameterNames.CODE_CHALLENGE, codeChallenge);
 			additionalParameters.put(PkceParameterNames.CODE_CHALLENGE_METHOD, "S256");
 		} catch (NoSuchAlgorithmException e) {
 			additionalParameters.put(PkceParameterNames.CODE_CHALLENGE, codeVerifier);
 		}
 	}
 
-	private String createCodeChallenge(String codeVerifier) throws NoSuchAlgorithmException {
+	private String createHash(String codeVerifier) throws NoSuchAlgorithmException {
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java]: Please rename `codeVerifier` to `value`

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/DefaultServerOAuth2AuthorizationRequestResolver.java [hunk 9]
```diff
@@ -52,6 +53,7 @@
  * used to resolve the {@link ClientRegistration} and create the {@link OAuth2AuthorizationRequest}.
  *
  * @author Rob Winch
+ * @author Mark Heckler
  * @since 5.1
  */
 public class DefaultServerOAuth2AuthorizationRequestResolver
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/DefaultServerOAuth2AuthorizationRequestResolver.java]: Please apply same changes as requested in `DefaultOAuth2AuthorizationRequestResolver`

File: oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/endpoint/OAuth2AuthorizationResponse.java [hunk 10]
```diff
@@ -63,6 +65,15 @@ public String getCode() {
 		return this.code;
 	}
 
+	/**
+	 * Returns the nonce.
+	 *
+	 * @return the nonce
+	 */
+	public String getNonce() {
```
Comment (jgrandja) [inline oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/endpoint/OAuth2AuthorizationResponse.java]: The `nonce` does not come back in the Authorization Response. It comes back in the ID Token. Please remove.

**Comment by mkheck** [comment]
Hi @jgrandja, 

Good catches one and all. I've reworked per your recommendations and all build tests pass. LDAP integration tests are failing, but I suppose those are resolved by Travis?

I've also formatted the commit message per the guidelines you provided. 😃 

Please let me know of any additional changes required or desired. Thanks!

**Comment by mkheck** [comment]
Missed 4 unit tests in `OAuth2AuthorizationRequestRedirectFilterTests`, fixed & re-pushed. Please let me know if I need to repackage in some way. 😑

Still getting local failures with various LDAP tests under `org.springframework.security.config.ldap`, but I did not modify those. Integration tests to complete successfully in CI pipeline?

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java [hunk 11]
```diff
@@ -201,6 +207,28 @@ private static String expandRedirectUri(HttpServletRequest request, ClientRegist
...
+	 * @see <a target="_blank" href="https://openid.net/specs/openid-connect-core-1_0.html#NonceNotes">15.5.2.  Nonce Implementation Notes</a>
+	 * @see <a target="_blank" href="https://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation">3.1.3.7.  ID Token Validation</a>
+	 */
+	private void addNonceParameters(Map<String, Object> attributes, Map<String, Object> additionalParameters) {
+		String nonce = this.stringKeyGenerator.generateKey();
+		attributes.put(IdTokenClaimNames.NONCE, nonce);
+
+		try {
+			String nonceHash = createHash(nonce);
+			additionalParameters.put(IdTokenClaimNames.NONCE, nonceHash);
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java]: This should be using `OidcParameterNames.NONCE`

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java [hunk 12]
```diff
@@ -201,6 +207,28 @@ private static String expandRedirectUri(HttpServletRequest request, ClientRegist
...
+		} catch (NoSuchAlgorithmException e) {
+			//
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java]: Looks like this [comment](https://github.com/spring-projects/spring-security/pull/7337#discussion_r322893893) was missed.

If the `nonceHash` could not be created than the `nonce` should not be added to the `attributes`. We should add a test for this as well.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizationCodeAuthenticationProvider.java [hunk 13]
```diff
@@ -75,6 +80,7 @@
 	private static final String INVALID_STATE_PARAMETER_ERROR_CODE = "invalid_state_parameter";
 	private static final String INVALID_REDIRECT_URI_PARAMETER_ERROR_CODE = "invalid_redirect_uri_parameter";
 	private static final String INVALID_ID_TOKEN_ERROR_CODE = "invalid_id_token";
+	private static final String INVALID_NONCE = "invalid_nonce";
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizationCodeAuthenticationProvider.java]: Please add `_ERROR_CODE` suffix just like other errors.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/DefaultServerOAuth2AuthorizationRequestResolver.java [hunk 14]
```diff
@@ -207,6 +211,28 @@ private static String expandRedirectUri(ServerHttpRequest request, ClientRegistr
...
+	 * Creates nonce and its hash for use in OpenID Connect Authentication Requests
+	 *
+	 * @param attributes where {@link IdTokenClaimNames#NONCE} is stored for the token request
+	 * @param additionalParameters where hash of {@link IdTokenClaimNames#NONCE} is added to the authentication request
+	 *
+	 * @since 5.2
+	 * @see <a target="_blank" href="https://openid.net/specs/openid-connect-core-1_0.html#NonceNotes">15.5.2.  Nonce Implementation Notes</a>
+	 * @see <a target="_blank" href="https://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation">3.1.3.7.  ID Token Validation</a>
+	 */
+	private void addNonceParameters(Map<String, Object> attributes, Map<String, Object> additionalParameters) {
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/DefaultServerOAuth2AuthorizationRequestResolver.java]: Please apply the same changes requested for `DefaultOAuth2AuthorizationRequestResolver`

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizationCodeAuthenticationProvider.java [hunk 15]
```diff
@@ -154,6 +160,22 @@ public Authentication authenticate(Authentication authentication) throws Authent
 		}
 		OidcIdToken idToken = createOidcToken(clientRegistration, accessTokenResponse);
 
+		String requestNonce = authorizationRequest.getAttribute(OidcParameterNames.NONCE);
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizationCodeAuthenticationProvider.java]: Looks like this same logic was missed in `OidcAuthorizationCodeReactiveAuthenticationManager`

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenValidatorTests.java [hunk 16]
```diff
@@ -26,11 +26,7 @@
 
 import java.time.Duration;
 import java.time.Instant;
-import java.util.Arrays;
-import java.util.Collection;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.Map;
+import java.util.*;
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenValidatorTests.java]: There are no changes in this file. Please reset.

File: oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/oidc/endpoint/OidcParameterNames.java [hunk 17]
```diff
@@ -20,6 +20,7 @@
  * and used by the authorization endpoint and token endpoint.
  *
  * @author Joe Grandja
+ * @author Mark Heckler
```
Comment (jgrandja) [inline oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/oidc/endpoint/OidcParameterNames.java]: Please update copyright year.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java (hunk 11 above)
Comment (mkheck) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java]: Sorry about that, corrected.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java (hunk 12 above)
Comment (mkheck) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java]: Corrected. I was following the PKCE approach suggested, and they add an attribute even if they're unable to create the hash...but they also add that same unencrypted value if hashing fails too. :/ At any rate, I've change to only add nonce attribute if we can also add the hash to additional parameters.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizationCodeAuthenticationProvider.java (hunk 13 above)
Comment (mkheck) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizationCodeAuthenticationProvider.java]: Done.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizationCodeAuthenticationProvider.java (hunk 15 above)
Comment (mkheck) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizationCodeAuthenticationProvider.java]: Indeed! Done now, thanks for catching that! 👍

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenValidatorTests.java (hunk 16 above)
Comment (mkheck) [reply oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenValidatorTests.java]: Done.

File: oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/oidc/endpoint/OidcParameterNames.java (hunk 17 above)
Comment (mkheck) [reply oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/oidc/endpoint/OidcParameterNames.java]: Done. Found another I'd missed & updated that too.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/DefaultServerOAuth2AuthorizationRequestResolver.java (hunk 14 above)
Comment (mkheck) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/DefaultServerOAuth2AuthorizationRequestResolver.java]: Done.

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizationCodeAuthenticationProviderTests.java [hunk 18]
```diff
@@ -302,6 +326,37 @@ public void authenticateWhenTokenSuccessResponseThenAdditionalParametersAddedToU
 				this.accessTokenResponse.getAdditionalParameters());
 	}
 
+	// gh-4442
+	@Test
+	public void authenticateWhenTokenSuccessResponseThenAdditionalParametersAddedToUserRequestNoNonce() {
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizationCodeAuthenticationProviderTests.java]: I'm not sure I understand the purpose for this test. I think we can remove this?

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java [hunk 19]
```diff
@@ -201,6 +207,27 @@ private static String expandRedirectUri(HttpServletRequest request, ClientRegist
...
+	 * @since 5.2
+	 * @see <a target="_blank" href="https://openid.net/specs/openid-connect-core-1_0.html#NonceNotes">15.5.2.  Nonce Implementation Notes</a>
+	 * @see <a target="_blank" href="https://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation">3.1.3.7.  ID Token Validation</a>
+	 */
+	private void addNonceParameters(Map<String, Object> attributes, Map<String, Object> additionalParameters) {
+		try {
+			String nonce = this.stringKeyGenerator.generateKey();
+			attributes.put(OidcParameterNames.NONCE, nonce);
+			
+			String nonceHash = createHash(nonce);
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/DefaultOAuth2AuthorizationRequestResolver.java]: If this errors than `attributes ` will still contain `OidcParameterNames.NONCE`.

I think you want this:

```
String nonce = this.stringKeyGenerator.generateKey();
String nonceHash = createHash(nonce);
additionalParameters.put(OidcParameterNames.NONCE, nonceHash);
attributes.put(OidcParameterNames.NONCE, nonce);
```

Also, please add a test for this scenario. The `attributes` and `additionalParameters` should always be in a consistent state.

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenValidatorTests.java (hunk 16 above)
Comment (jgrandja) [reply oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenValidatorTests.java]: Looks like this was missed

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenDecoderFactory.java [hunk 20]
```diff
@@ -88,12 +89,14 @@
...
 		Converter<Object, ?> instantConverter = getConverter(TypeDescriptor.valueOf(Instant.class));
 		Converter<Object, ?> urlConverter = getConverter(TypeDescriptor.valueOf(URL.class));
+		Converter<Object, ?> stringConverter = getConverter(TypeDescriptor.valueOf(String.class));
 		Converter<Object, ?> collectionStringConverter = getConverter(
 				TypeDescriptor.collection(Collection.class, TypeDescriptor.valueOf(String.class)));
 
 		Map<String, Converter<Object, ?>> claimTypeConverters = new HashMap<>();
 		claimTypeConverters.put(IdTokenClaimNames.ISS, urlConverter);
 		claimTypeConverters.put(IdTokenClaimNames.AUD, collectionStringConverter);
+		claimTypeConverters.put(IdTokenClaimNames.NONCE, stringConverter);
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcIdTokenDecoderFactory.java]: Please update test `OidcIdTokenDecoderFactoryTests.createDefaultClaimTypeConvertersWhenCalledThenDefaultsAreCorrect()`

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/ReactiveOidcIdTokenDecoderFactory.java [hunk 21]
```diff
@@ -88,12 +89,14 @@
...
 		Converter<Object, ?> instantConverter = getConverter(TypeDescriptor.valueOf(Instant.class));
 		Converter<Object, ?> urlConverter = getConverter(TypeDescriptor.valueOf(URL.class));
+		Converter<Object, ?> stringConverter = getConverter(TypeDescriptor.valueOf(String.class));
 		Converter<Object, ?> collectionStringConverter = getConverter(
 				TypeDescriptor.collection(Collection.class, TypeDescriptor.valueOf(String.class)));
 
 		Map<String, Converter<Object, ?>> claimTypeConverters = new HashMap<>();
 		claimTypeConverters.put(IdTokenClaimNames.ISS, urlConverter);
 		claimTypeConverters.put(IdTokenClaimNames.AUD, collectionStringConverter);
+		claimTypeConverters.put(IdTokenClaimNames.NONCE, stringConverter);
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/ReactiveOidcIdTokenDecoderFactory.java]: Please update test `ReactiveOidcIdTokenDecoderFactoryTests.createDefaultClaimTypeConvertersWhenCalledThenDefaultsAreCorrect()`

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizationCodeAuthenticationProviderTests.java [hunk 22]
```diff
@@ -328,4 +384,25 @@ private OAuth2AccessTokenResponse accessTokenSuccessResponse() {
 				.build();
 
 	}
+
+	/**
```
Comment (jgrandja) [inline oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizationCodeAuthenticationProviderTests.java]: We can remove javadoc

File: oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/oidc/endpoint/OidcParameterNames.java [hunk 23]
```diff
@@ -30,4 +31,9 @@
 	 */
 	String ID_TOKEN = "id_token";
 
+	/**
+	 * {@code nonce} - used in the Access Token Request and Response.
```
Comment (jgrandja) [inline oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/oidc/endpoint/OidcParameterNames.java]: javadoc should be "used in the Authentication Request"

**Comment by jgrandja** [comment]
Merged via da9f027
